testing/test_pathlib.py

```diff
@@ -615,3 +620,44 @@ def test_parent_contains_child_module_attribute(
         assert sorted(modules) == ["xxx", "xxx.tests", "xxx.tests.foo"]
         assert modules["xxx"].tests is modules["xxx.tests"]
         assert modules["xxx.tests"].foo is modules["xxx.tests.foo"]
+
+    def test_importlib_package(self, monkeypatch: MonkeyPatch, tmp_path: Path):
+        """
+        Importing a package using --importmode=importlib should not import the
+        package's __init__.py file more than once (#11306).
+        """
+        monkeypatch.chdir(tmp_path)
+        monkeypatch.syspath_prepend(tmp_path)
+
+        package_name = "importlib_import_package"
+        tmp_path.joinpath(package_name).mkdir()
+        init = tmp_path.joinpath(f"{package_name}/__init__.py")
+        init.write_text(
+            dedent(
+                """
+                from .singleton import Singleton
+
+                instance = Singleton()
+                """
+            ),
+            encoding="ascii",
+        )
+        singleton = tmp_path.joinpath(f"{package_name}/singleton.py")
+        singleton.write_text(
+            dedent(
+                """
+                class Singleton:
```

Comment:
i believe this could be simplified to a function + a module global

Reply:
Simplified to this:

```diff
diff --git a/testing/test_pathlib.py b/testing/test_pathlib.py
index 0878eec0e..848d265a7 100644
--- a/testing/test_pathlib.py
+++ b/testing/test_pathlib.py
@@ -635,9 +635,9 @@ class TestImportLibMode:
         init.write_text(
             dedent(
                 """
-                from .singleton import Singleton
+                from .singleton import initialize
 
-                instance = Singleton()
+                BOOM = initialize()
                 """
             ),
             encoding="ascii",
@@ -646,18 +646,17 @@ class TestImportLibMode:
         singleton.write_text(
             dedent(
                 """
-                class Singleton:
-                    BOOM = []
-
-                    def __init__(self) -> None:
-                        cls = type(self)
-                        cls.BOOM.append(True)
-                        if len(cls.BOOM) > 1:
-                            raise RuntimeError("Already initialized")
+                BOOM = []
+                
+                def initialize():
+                    BOOM.append(True)
+                    if len(BOOM) > 1:
+                        raise RuntimeError("Already initialized")
+                    return BOOM
                 """
             ),
             encoding="ascii",
         )
 
         mod = import_path(init, root=tmp_path, mode=ImportMode.importlib)
-        assert mod.instance.BOOM == [True]
+        assert mod.BOOM == [True]
```

However I think the `Singleton` example is a bit easier to understand, because the singleton pattern is usually known; this function and global variable is a bit contrived I think.

Reply:
Improved the example a bit, still keeping the singleton theme. Let me know what you think. 👍

Reply:
@RonnyPfannschmidt

Reply:
Looking better, if the check happens before the append, a boolean check can be used

Reply:
I see, but I personally like adding to a list, because it lets us directly test the number of times the singleton has been created. 👍